### PR TITLE
build: place NAPI-related changes behind a configure flag

### PR DESCRIPTION
**Pull request by gabrielschulhof**
Perhaps we should update the blurb at the [top of src/node_jsvmapi.cc](https://github.com/nodejs/abi-stable-node/blob/api-prototype-8.x/src/node_jsvmapi.cc#L1-L15).

### CODE REVIEW

File: src/node.cc [hunk 1]
```diff
@@ -2418,6 +2419,7 @@ void DLOpen(const FunctionCallbackInfo<Value>& args) {
     return;
   }
 
+#ifdef ENABLE_NAPI
 // try loading with node 0.10 manner in case of node 0.10
```
Comment (jasongin) [inline src/node.cc]: Why is support for old modules conditional on ENABLE_NAPI?

File: src/node_jsvmapi.h [hunk 2]
```diff
@@ -18,6 +18,7 @@
 #ifndef SRC_NODE_JSVMAPI_H_
 #define SRC_NODE_JSVMAPI_H_
 
+#include "node.h"
```
Comment (jasongin) [inline src/node_jsvmapi.h]: Can inclusion of `node.h` be avoided? It would be better if all the node APIs (and V8 APIs included by `node.h`) were not available when building a NAPI module.

File: src/node_jsvmapi.h [hunk 3]
```diff
@@ -35,6 +36,15 @@
 #endif
 
 namespace node {
+
+struct node_module_old {
```
Comment (jasongin) [inline src/node_jsvmapi.h]: Why is this moved here? Can it go in `node_jsvmapi_internal.h` instead?

File: src/node_jsvmapi.h (hunk 2 above)
Comment (mhdawson) [reply src/node_jsvmapi.h]: It would be better if modules needed to include that separately.

File: src/node_jsvmapi.h (hunk 3 above)
Comment (mhdawson) [reply src/node_jsvmapi.h]: I think it should be able to be in  node_jsvmapi_internal.h.  The only place the external definition should have been necessary was when building against 0.10.x and in that case it would not have been the "old" version but the version in node.h.

File: src/node_jsvmapi.h (hunk 2 above)
Comment (mhdawson) [reply src/node_jsvmapi.h]: Having said that I think that all modules will have to include node.h as they need the node_module definition which it contains.

Comment (mhdawson) [reply src/node_jsvmapi.h]: As a follow on we may want to refactor node.h so that is not the case because as Jason says we want to avoid leakage of the v8 apis into modules building against the NAPI api.

File: src/node.cc (hunk 1 above)
Comment (mhdawson) [reply src/node.cc]: node_module_old is a structure which reflects the structure of node_module in the 0.10.x release.  We only need it if we are going to try and load a module compiled against 0.10.x which only makes sense with NAPI (without NAPI you have to recompile your module to use it with newer node versions).

File: src/node.cc [hunk 4]
```diff
@@ -2475,16 +2477,21 @@ void DLOpen(const FunctionCallbackInfo<Value>& args) {
...
     uv_dlclose(&lib);
     env->ThrowError("Module did not self-register.");
     return;
   }
 
+#ifdef ENABLE_NAPI
   bool isNapiModule = mp->nm_version == -1;
 
   if (mp->nm_version != NODE_MODULE_VERSION && !isNapiModule) {
+#else /* !defined ENABLE_NAPI */
```
Comment (mhdawson) [inline src/node.cc]: This is a case were if we get agreement to enable NAPI by default at compile time, we may need a runtime flag to make sure we don't try to load modules were the version is -1 unless the runtime Flag has enabled NAPI

File: test/addons-abi/test_object/test_object.cc [hunk 5]
```diff
@@ -1,4 +1,4 @@
-﻿#include <node_jsvmapi.h>
+#include <node_jsvmapi.h>
```
Comment (mhdawson) [inline test/addons-abi/test_object/test_object.cc:1]: There are few changes like this that I'm not sure what actually changed ?  Is it whitespace ?

File: src/node_jsvmapi.h (hunk 2 above)
Comment (jasongin) [reply src/node_jsvmapi.h]: We got by without any inclusion of `node.h` before. It works because `node_jsvmapi.h` defines its own `NODE_MODULE_ABI` macro that is used for module registration. I don't think any refactoring of `node.h` is necessary.

File: test/addons-abi/test_object/test_object.cc (hunk 5 above)
Comment (gabrielschulhof) [reply test/addons-abi/test_object/test_object.cc:1]:
```diff
-<U+FEFF>#include <node_jsvmapi.h>
+#include <node_jsvmapi.h>
```

File: src/node_jsvmapi.h (hunk 2 above)
Comment (mhdawson) [reply src/node_jsvmapi.h]: ah right, good.

**Comment by gabrielschulhof** [comment]
To keep a reference to the 0.10 stuff I could point a branch to the commit preceding the one that lands these changes.

**Comment by mhdawson** [comment]
One more thought is that we may want to make the default to "on" in the abi-stable-node repo and then change that to off if necessary when we PR to node core.

**Comment by gabrielschulhof** [comment]
@mhdawson roger that! Can do.

**Review by mhdawson — APPROVED** [review]
LGTM
